### PR TITLE
front: align compute margin to spec in timestops table

### LINKED ISSUE #8959: Margin bugs and input row deletion bugs
### What happened?

I.
Margin of `0min/100km` still computes margin

image

II.
No margin defined (input table):
image

But still, some margins are somehow displayed (output table):
image

III.
Input departure time is bugged when try to update an existing value

https://github.com/user-attachments/assets/fdbf2d51-40d1-4ffe-9405-1d462d3c2f47

IV.
Multi selected fields modification is bugged

https://github.com/user-attachments/assets/74c09e44-5942-4329-967c-8b7c26090056

V.
Delete the value of a stopping time field behaves weirdly
(cf. previous video)

### What did you expect to happen?

_No response_

### How can we reproduce it (as minimally and precisely as possible)?

1. open input table
2. play with it
3. open output table

### On which environments the bug occurs?

Recette (SNCF)

### On which browser the bug occurs?

Firefox

### OSRD version (top right corner `Account` button > `Informations`)

staging fbf8e3e

### LINKED ISSUE #9737: Improve Margin Display in Timestop Tables
### Description and goal

Enhance the clarity and functionality of margin displays in the input and output table, and enforce consistency in which margins we choose to display.

### Acceptance criteria

- [x] Input Table:

    Display all user-specified margins, including default and zero values, as well as successive identical values.

- [x] Output table:

  All margins displayed as outlined in https://github.com/osrd-project/osrd-confidential/issues/852 and https://sncf.sharepoint.com/:p:/r/sites/OSRD644GrpO365/Documents%20partages/PO/Tableau%20input-output/Cas%20marges.pptx?d=w532257c9e47f47429bb557a71d866740&csf=1&web=1&e=lbWEM8

### PR DESCRIPTION
**Pull request by Synar**
Close #9737
Also partially fix #8959 

Depend on https://github.com/OpenRailAssociation/osrd/pull/9642

### CODE REVIEW

File: front/src/modules/timesStops/helpers/computeMargins.ts [hunk 1]
```diff
@@ -1,41 +1,53 @@
...
+  TrainScheduleWithDetails,
+} from 'modules/trainschedule/components/Timetable/types';
 import { ms2sec } from 'utils/timeManipulation';
 
 import { formatDigitsAndUnit } from './utils';
+import type { ScheduleEntry } from '../types';
 
-function getTheoreticalMargin(selectedTrainSchedule: TrainScheduleResult, pathStepId: string) {
-  if (selectedTrainSchedule.path.length === 0) {
+export function getTheoreticalMargins(selectedTrainSchedule: TrainScheduleResult) {
```
Comment (clarani) [inline front/src/modules/timesStops/helpers/computeMargins.ts:10]: Nice function !
- can you add a comment to describe in a sentence what this function does ?
- can you add some tests for this small functions ?

File: front/src/modules/timesStops/helpers/computeMargins.ts [hunk 2]
```diff
@@ -1,41 +1,53 @@
...
+  const selectedTrainScheduleMargins = selectedTrainSchedule.margins;
```
Comment (clarani) [inline front/src/modules/timesStops/helpers/computeMargins.ts]: You can extract directly margins like this
```suggestion
  const { margins } = selectedTrainSchedule;
```

File: front/src/modules/timesStops/helpers/computeMargins.ts [hunk 3]
```diff
@@ -1,41 +1,53 @@
...
-    (margins.values.length === 1 && margins.values[0] === '0%') ||
-    pathStepIndex === selectedTrainSchedule.path.length - 1
+    pathStepIndex === selectedTrainSchedule.path.length - 1 ||
+    !theoreticalMargins ||
+    !theoreticalMargins[pathStepId] ||
+    !(
+      scheduleByAt[pathStepId]?.arrival ||
+      scheduleByAt[pathStepId]?.stop_for ||
+      theoreticalMargins[pathStepId].isBoundary
+    )
```
Comment (clarani) [inline front/src/modules/timesStops/helpers/computeMargins.ts]: I think you can simplify like this:
```suggestion
  const schedule = scheduleByAt[pathStepId];
  const currentTheoreticalMargin = theoreticalMargins?.[pathStepId];
  if (pathStepIndex === path.length - 1 || !currentTheoreticalMargin || !(schedule || currentTheoreticalMargin?.isBoundary)
  ) {
    return {
      theoreticalMargin: undefined,
      theoreticalMarginSeconds: undefined,
      calculatedMargin: undefined,
      diffMargins: undefined,
    };
  }
```

We don't display the margins if:
- the pathStep is the final pathStep
- the trainSchedule has no margins (and no theoreticalMargins and no currentTheoreticalMargin)
- the pathStep has no constraint:
  - either no schedule (if it has a schedule, it should have either an arrival or a stop_for)
  - or if no new margin is defined at this pathStep

File: front/src/modules/timesStops/helpers/computeMargins.ts [hunk 4]
```diff
@@ -45,14 +57,21 @@ function computeMargins(
...
-  for (let index = 1; index < pathStepIndex; index += 1) {
-    if (margins.boundaries.includes(path[index].id)) {
-      prevIndex = index;
+  for (let index = pathStepIndex + 1; index < path.length; index += 1) {
+    const curStepId = path[index].id;
+    const curStepSchedule = scheduleByAt[curStepId];
+    if (
+      theoreticalMargins[curStepId]?.isBoundary ||
+      curStepSchedule?.arrival ||
+      curStepSchedule?.stop_for
```
Comment (clarani) [inline front/src/modules/timesStops/helpers/computeMargins.ts]: checking if `curStepSchedule` is defined is enough
```suggestion
      curStepSchedule
```

File: front/src/modules/timesStops/helpers/utils.ts [hunk 5]
```diff
@@ -114,7 +114,7 @@ const getDigits = (unit: string | undefined) =>
   unit === MarginUnit.second || unit === MarginUnit.percent ? 0 : 1;
 
 export function formatDigitsAndUnit(fullValue: string | number | undefined, unit?: string) {
-  if (fullValue === undefined || fullValue === '0%') {
+  if (fullValue === undefined) {
```
Comment (clarani) [inline front/src/modules/timesStops/helpers/utils.ts:117]: Why did you remove the "0%" here ?

File: front/src/modules/trainschedule/components/Timetable/types.ts [hunk 6]
```diff
@@ -43,3 +43,8 @@ export type InvalidReason =
   | Extract<SimulationSummaryResult['status'], 'pathfinding_failure' | 'simulation_failed'>
   | PathfindingNotFound['error_type']
   | PathfindingInputError['error_type'];
+
+export type TheoreticalMarginsRecord = Record<
+  string,
+  { theoreticalMargin: string; isBoundary: boolean }
+>;
```
Comment (clarani) [inline front/src/modules/trainschedule/components/Timetable/types.ts]: You can move this type in the types of `modules/timesStops` since it used only there

File: front/src/modules/timesStops/types.ts
Comment (clarani) [inline front/src/modules/timesStops/types.ts:1]: Nice comment :)

**Review by louisgreiner — COMMENTED** [review]
Is this behavior normal ? Theoretical margin is bold for some PRs, but not bold for the rest of the PRs

image

image

Edit: yes it is ->
- bold : when user has forced `0%` of margin
- not bold : when the user did not set margin, but the PR is especially asked for the pathfinding

File: front/src/modules/timesStops/hooks/useTimeStopsColumns.tsx [hunk 7]
```diff
@@ -167,12 +169,34 @@ export const useTimeStopsColumns = <T extends TimeStopsRow>(
             alignRight: true,
           })
         ),
+        ...(isOutputTable && {
+          component: ({ rowData }) => {
+            const [digits, unit] = rowData.theoreticalMargin?.split(NO_BREAK_SPACE) ?? [];
```
Comment (clarani) [inline front/src/modules/timesStops/hooks/useTimeStopsColumns.tsx]: Can you add an early return if `theoreticalMargin` is undefined ? You can return null

File: front/src/modules/timesStops/styles/_timesStopsDatasheet.scss [hunk 8]
```diff
@@ -19,8 +23,20 @@
...
-  .invalidCell {
-    color: red !important;
+  .self-center {
+    align-self: center;
+  }
+
+  .small-unit-container {
+    display: inline-block;
+    font-size: 0.5rem;
+    line-height: 0.5rem;
```
Comment (clarani) [inline front/src/modules/timesStops/styles/_timesStopsDatasheet.scss]: Can you use px for line-height for the moment ? :) Maybe it's not very common, but it's our guideline for the moment

**Review by louisgreiner — APPROVED** [review]
Tested and conform to the specs 😄

File: front/src/modules/timesStops/helpers/computeMargins.ts (hunk 1 above)
Comment (Synar) [reply front/src/modules/timesStops/helpers/computeMargins.ts:10]: Done thanks!

File: front/src/modules/timesStops/helpers/computeMargins.ts (hunk 2 above)
Comment (Synar) [reply front/src/modules/timesStops/helpers/computeMargins.ts]: Sure

File: front/src/modules/timesStops/helpers/computeMargins.ts (hunk 3 above)
Comment (Synar) [reply front/src/modules/timesStops/helpers/computeMargins.ts]: Great suggestion thanks!

(I renamed currentTheoreticalMargin into stepTheoreticalMarginInfo to avoid confusion with other variables that follow)

File: front/src/modules/timesStops/helpers/computeMargins.ts (hunk 4 above)
Comment (Synar) [reply front/src/modules/timesStops/helpers/computeMargins.ts]: Thanks!

File: front/src/modules/timesStops/helpers/utils.ts (hunk 5 above)
Comment (Synar) [reply front/src/modules/timesStops/helpers/utils.ts:117]: Because according to the last spec, I believe we want to display all margin values, even if they are equal to 0%

File: front/src/modules/trainschedule/components/Timetable/types.ts (hunk 6 above)
Comment (Synar) [reply front/src/modules/trainschedule/components/Timetable/types.ts]: Oops, I confused timetable and timesstops, thanks!

File: front/src/modules/timesStops/hooks/useTimeStopsColumns.tsx (hunk 7 above)
Comment (Synar) [reply front/src/modules/timesStops/hooks/useTimeStopsColumns.tsx]: Sure!

File: front/src/modules/timesStops/styles/_timesStopsDatasheet.scss (hunk 8 above)
Comment (Synar) [reply front/src/modules/timesStops/styles/_timesStopsDatasheet.scss]: Sure! I  used rems because I thought line-height was "font-size based", Thibaut initially asked for 8px for both ^^.

File: front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json [hunk 9]
```diff
@@ -25,9 +25,9 @@
     "shortSlipDistance": true,
     "margin": {
       "theoretical": "1 min/100km",
-      "theoreticalS": "37 s",
-      "actual": "1213 s",
-      "difference": "1176 s"
+      "theoreticalS": "9 s",
+      "actual": "291 s",
+      "difference": "283 s"
```
Comment (Maymanaf) [inline front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: Can the margin calculation be optimized to ensure an accurate margin difference and eliminate the ±1 second discrepancy?

Comment (Synar) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: What +/- 1 second discrepancy do you mean?

Comment (Maymanaf) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: The difference here should be 282s

Comment (Synar) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: Oh right! I'll check

File: front/src/modules/timesStops/styles/_timesStopsDatasheet.scss (hunk 8 above)
Comment (clarani) [reply front/src/modules/timesStops/styles/_timesStopsDatasheet.scss]: Yes I know that line-height is based on font-size :/ Maybe we should also use rem for line-height, we can discuss this in our weekly meeting

Comment (Synar) [reply front/src/modules/timesStops/styles/_timesStopsDatasheet.scss]: Sure! I did put it in px in the meantime

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:

File: front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json (hunk 9 above)
Comment (Synar) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: So the issue is that we round the difference of the margins after doing the difference. So there can be a +/-1 second discrepancy between "round(thmargin - cmargin)" and "round(thmargin) - round(cmargin)"

We could instead do the later, which would solve the discrepancy, but would be less accurate to the raw difference we get from core

What do you think?

Comment (Maymanaf) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: We know that our calculation are correct but the visuel display should accurate for the user.
I think we should go for `(round(thmargin) - round(cmargin) `

Comment (Maymanaf) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: @clarani can you confirm

Comment (Synar) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: I asked maelys, and she confirms we want round(th)-round(cm), I'll make the change asap, thanks for the catch!

Comment (Synar) [reply front/tests/assets/operationStudies/timesAndStops/expectedOutputsCellsData.json]: Done!

**Review by Maymanaf — APPROVED** [review]
LGTM for e2e tests ✅
